core/src/main/scala/kafka/server/ClientMetricsManager.scala

```diff
@@ -0,0 +1,93 @@
+import org.apache.kafka.common.Uuid
+import org.apache.kafka.common.message.GetTelemetrySubscriptionsResponseData
+import org.apache.kafka.common.protocol.Errors
+import org.apache.kafka.common.record.CompressionType
+import org.apache.kafka.common.requests.{GetTelemetrySubscriptionRequest, GetTelemetrySubscriptionResponse}
+
+import java.util.Properties
+import scala.collection.mutable.ListBuffer
+import scala.jdk.CollectionConverters._
+
+object ClientMetricsManager {
+  val CM_CACHE_GC_INTERVAL = 5 * 60 * 1000 // 5 minutes
+  val CM_CACHE_MAX_SIZE = 1024
+
+  private val _instance = new ClientMetricsManager
+  def getInstance = _instance
+
+  def getSupportedCompressionTypes: List[java.lang.Byte] = {
+    val compressionTypes = new ListBuffer[java.lang.Byte]
+    CompressionType.values.foreach(x => compressionTypes.append(x.id.toByte))
+    compressionTypes.toList
+  }
+
+  def processGetSubscriptionRequest(request: RequestChannel.Request,
+                                    throttleMs: Int): GetTelemetrySubscriptionResponse = {
+    val subscriptionRequest = request.body[GetTelemetrySubscriptionRequest]
+    val clientInfo = CmClientInformation(request, subscriptionRequest.getClientInstanceId.toString)
+    _instance.processGetSubscriptionRequest(subscriptionRequest, clientInfo, throttleMs)
+  }
+
+}
+
+class ClientMetricsManager {
+  val clientInstanceCache = ClientMetricsCache.getInstance
+  def getCacheSize = clientInstanceCache.getSize
+  def clearCache() = clientInstanceCache.clear()
+  def getClientInstance(id: Uuid) = clientInstanceCache.get(id)
+
+  def processGetSubscriptionRequest(subscriptionRequest: GetTelemetrySubscriptionRequest,
+                                    clientInfo: CmClientInformation,
+                                    throttleMs: Int): GetTelemetrySubscriptionResponse = {
+    var clientInstanceId = subscriptionRequest.getClientInstanceId
+    if (clientInstanceId == null || clientInstanceId == Uuid.ZERO_UUID) {
+      clientInstanceId = Uuid.randomUuid()
+    }
+    var clientInstance = getClientInstance(clientInstanceId)
+    if (clientInstance == null) {
+      clientInstance = createClientInstance(clientInstanceId, clientInfo)
+    }
+
+    val data =  new GetTelemetrySubscriptionsResponseData()
+        .setThrottleTimeMs(throttleMs)
+        .setClientInstanceId(clientInstanceId)
+        .setSubscriptionId(clientInstance.getSubscriptionId) // TODO: should we use LONG instead of int?
+        .setAcceptedCompressionTypes(getSupportedCompressionTypes.asJava)
+        .setPushIntervalMs(clientInstance.getPushIntervalMs)
+        .setDeltaTemporality(true)
+        .setErrorCode(Errors.NONE.code())
+        .setRequestedMetrics(clientInstance.getMetrics.asJava)
+
+    if (clientInstance.isDisabledForMetricsCollection) {
+      info(s"Metrics collection is disabled for the client: ${clientInstance.getId.toString}")
+      data.setErrorCode(Errors.INVALID_CONFIG.code)
```

Comment:
This case should not return an error to the client, just an empty (thus disabled) subscription set.

Reply:
Agreed

Reply:
Done